fedimint-core/src/net/iroh.rs

```diff
@@ -0,0 +1,95 @@
+use std::net::SocketAddr;
+
+use fedimint_core::util::SafeUrl;
+use fedimint_logging::LOG_NET_IROH;
+use iroh::defaults::DEFAULT_STUN_PORT;
+use iroh::discovery::pkarr::{PkarrPublisher, PkarrResolver};
+use iroh::{Endpoint, RelayMode, RelayNode, RelayUrl, SecretKey};
+use iroh_relay::RelayQuicConfig;
+use tracing::info;
+use url::Url;
+
+use crate::iroh_prod::{FM_IROH_DNS_FEDIMINT_PROD, FM_IROH_RELAYS_FEDIMINT_PROD};
+
+pub async fn build_iroh_endpoint(
+    secret_key: SecretKey,
+    bind_addr: SocketAddr,
+    iroh_dns: Option<SafeUrl>,
+    iroh_relays: Vec<SafeUrl>,
+    alpn: &[u8],
+) -> Result<Endpoint, anyhow::Error> {
+    let iroh_dns_servers: Vec<_> = iroh_dns.clone().map_or_else(
+        || {
+            FM_IROH_DNS_FEDIMINT_PROD
+                .into_iter()
+                .map(|dns| dns.parse().expect("Can't fail"))
+                .collect()
+        },
+        |iroh_dns| vec![iroh_dns.to_unsafe()],
+    );
+
+    let relay_mode = if iroh_relays.is_empty() {
+        RelayMode::Custom(
+            FM_IROH_RELAYS_FEDIMINT_PROD
+                .into_iter()
+                .map(|url| RelayNode {
+                    url: RelayUrl::from(Url::parse(url).expect("Hardcoded, can't fail")),
+                    stun_only: false,
+                    stun_port: DEFAULT_STUN_PORT,
+                    quic: Some(RelayQuicConfig::default()),
+                })
+                .collect(),
+        )
+    } else {
+        RelayMode::Custom(
+            iroh_relays
+                .into_iter()
+                .map(|url| RelayNode {
+                    url: RelayUrl::from(url.to_unsafe()),
+                    stun_only: false,
+                    stun_port: DEFAULT_STUN_PORT,
+                    quic: Some(RelayQuicConfig::default()),
+                })
+                .collect(),
+        )
+    };
+
+    let mut builder = Endpoint::builder();
+
+    for iroh_dns in iroh_dns_servers {
+        builder = builder
+            .add_discovery({
```

Comment:
Pkarr in stable version has so many issues, that I would recommend adding a toggle (with env var switch) to disable it, right away. Check how both client and server do it now.